File: src/lib/settingsWindow.ts
import { isTauriEnv } from './tauri-bridge';

export const SETTINGS_WINDOW_LABEL = 'settings';
export const SETTINGS_OPEN_EVENT = 'settings://open';
export const SETTINGS_TABS = ['appearance', 'shortcuts', 'supervisor', 'data'] as const;

export type SettingsTab = (typeof SETTINGS_TABS)[number];

export interface SettingsOpenPayload {
  tab: SettingsTab;
}

export interface SettingsWindowCreateOptions {
  url: string;
  title: string;
  width: number;
  height: number;
  minWidth: number;
  minHeight: number;
  resizable: boolean;
  decorations: boolean;
  transparent: boolean;
  center: boolean;
  visible: boolean;
  focus: boolean;
  skipTaskbar: boolean;
}

export interface SettingsWindowHandle {
  show: () => Promise<void>;
  unminimize: () => Promise<void>;
  setFocus: () => Promise<void>;
  once: <T = unknown>(
    event: string,
    handler: (event: { payload: T }) => void,
  ) => Promise<() => void>;
}

export interface SettingsWindowAdapters {
  getExistingWindow: () => Promise<SettingsWindowHandle | null>;
  createWindow: (options: SettingsWindowCreateOptions) => SettingsWindowHandle;
  emitTo: (target: string, event: string, payload: SettingsOpenPayload) => Promise<void>;
  openBrowserWindow?: (url: string) => boolean;
}

export const SETTINGS_WINDOW_BASE_OPTIONS = {
  title: 'ThreadTerm Settings',
  width: 960,
  height: 720,
  minWidth: 760,
  minHeight: 520,
  resizable: true,
  decorations: true,
  transparent: false,
  center: true,
  visible: true,
  focus: true,
  skipTaskbar: false,
} as const;

export function normalizeSettingsTab(tab: unknown): SettingsTab {
  return SETTINGS_TABS.includes(tab as SettingsTab) ? (tab as SettingsTab) : 'shortcuts';
}

export function buildSettingsWindowUrl(tab: unknown): string {
  return `settings.html?tab=${encodeURIComponent(normalizeSettingsTab(tab))}`;
}

export function normalizeSettingsOpenPayload(payload: unknown): SettingsOpenPayload {
  if (typeof payload === 'object' && payload !== null && 'tab' in payload) {
    return { tab: normalizeSettingsTab((payload as { tab?: unknown }).tab) };
  }
  return { tab: normalizeSettingsTab(payload) };
}

async function focusSettingsWindow(windowHandle: SettingsWindowHandle): Promise<void> {
  await windowHandle.show().catch(() => {});
  await windowHandle.unminimize().catch(() => {});
  await windowHandle.setFocus().catch(() => {});
}

export async function openSettingsWindowWithAdapters(
  tabInput: unknown,
  adapters: SettingsWindowAdapters,
): Promise<boolean> {
  const tab = normalizeSettingsTab(tabInput);
  const existing = await adapters.getExistingWindow();

  if (existing) {
    await adapters.emitTo(SETTINGS_WINDOW_LABEL, SETTINGS_OPEN_EVENT, { tab }).catch((error) => {
      console.debug('[settings] failed to route settings tab', error);
    });
    await focusSettingsWindow(existing);
    return true;
  }

  const options: SettingsWindowCreateOptions = {
    ...SETTINGS_WINDOW_BASE_OPTIONS,
    url: buildSettingsWindowUrl(tab),
  };

  try {
    const created = adapters.createWindow(options);
    return await new Promise<boolean>((resolve) => {
      let settled = false;
      const settle = (value: boolean) => {
        if (settled) return;
        settled = true;
        resolve(value);
      };

      void created.once('tauri://created', () => {
        void focusSettingsWindow(created).finally(() => settle(true));
      });
      void created.once('tauri://error', (event) => {
        console.warn('[settings] failed to create settings window', event.payload);
        settle(false);
      });
    });
  } catch (error) {
    console.warn('[settings] failed to create settings window', error);
    return false;
  }
}

function openBrowserSettingsWindow(tab: unknown): boolean {
  if (typeof window === 'undefined') return false;
  const opened = window.open(
    buildSettingsWindowUrl(tab),
    SETTINGS_WINDOW_LABEL,
    'width=960,height=720,resizable=yes,scrollbars=yes',
  );
  opened?.focus?.();
  return Boolean(opened);
}

async function getTauriSettingsWindowAdapters(): Promise<SettingsWindowAdapters> {
  const [{ WebviewWindow }, { emitTo }] = await Promise.all([
    import('@tauri-apps/api/webviewWindow'),
    import('@tauri-apps/api/event'),
  ]);

  return {
    getExistingWindow: () =>
      WebviewWindow.getByLabel(SETTINGS_WINDOW_LABEL) as Promise<SettingsWindowHandle | null>,
    createWindow: (options) =>
      new WebviewWindow(SETTINGS_WINDOW_LABEL, options) as SettingsWindowHandle,
    emitTo: (target, event, payload) => emitTo(target, event, payload),
  };
}

export async function openSettingsWindow(tab: unknown = 'shortcuts'): Promise<boolean> {
  if (!isTauriEnv()) {
    return openBrowserSettingsWindow(tab);
  }

  const adapters = await getTauriSettingsWindowAdapters();
  return openSettingsWindowWithAdapters(tab, adapters);
}

File: src-tauri/src/settings_window.rs
use serde::Serialize;
use tauri::{Emitter, Manager, WebviewUrl, WebviewWindowBuilder};

const SETTINGS_WINDOW_LABEL: &str = "settings";
const SETTINGS_OPEN_EVENT: &str = "settings://open";

#[derive(Clone, Serialize)]
struct SettingsOpenPayload {
    tab: &'static str,
}

fn normalize_settings_tab(value: &str) -> &'static str {
    match value {
        "appearance" => "appearance",
        "supervisor" => "supervisor",
        "data" => "data",
        _ => "shortcuts",
    }
}

fn focus_settings_window(window: &tauri::WebviewWindow) {
    let _ = window.show();
    let _ = window.unminimize();
    let _ = window.set_focus();
}

#[tauri::command]
pub async fn settings_open(app: tauri::AppHandle, tab: String) -> Result<(), String> {
    let tab = normalize_settings_tab(&tab);
    if let Some(window) = app.get_webview_window(SETTINGS_WINDOW_LABEL) {
        app.emit_to(
            SETTINGS_WINDOW_LABEL,
            SETTINGS_OPEN_EVENT,
            SettingsOpenPayload { tab },
        )
        .map_err(|error| format!("Could not route the settings tab: {error}"))?;
        focus_settings_window(&window);
        return Ok(());
    }

    let builder = WebviewWindowBuilder::new(
        &app,
        SETTINGS_WINDOW_LABEL,
        WebviewUrl::App(format!("settings.html?tab={tab}").into()),
    )
    .title("ThreadTerm Settings")
    .inner_size(960.0, 720.0)
    .min_inner_size(760.0, 520.0)
    .resizable(true)
    .decorations(true)
    .transparent(false)
    .center()
    .visible(true)
    .focused(true)
    .skip_taskbar(false);
    let builder = crate::data_directory::apply_webview_data_directory(&app, builder);
    let window = builder
        .build()
        .map_err(|error| format!("Could not create the settings window: {error}"))?;
    focus_settings_window(&window);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::normalize_settings_tab;

    #[test]
    fn only_accepts_known_settings_tabs() {
        assert_eq!(normalize_settings_tab("appearance"), "appearance");
        assert_eq!(normalize_settings_tab("supervisor"), "supervisor");
        assert_eq!(normalize_settings_tab("data"), "data");
        assert_eq!(normalize_settings_tab("missing?injected=true"), "shortcuts");
    }
}
